File: sdks/go/relayhub/transport.go
package relayhub

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"math/rand"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"
)

var retryableStatus = map[int]bool{429: true, 500: true, 502: true, 503: true, 504: true}

// requestOptions carries per-call overrides. Zero value means "use client defaults."
type requestOptions struct {
	query          url.Values
	headers        map[string]string
	timeout        time.Duration
	maxRetries     *int
	idempotencyKey string
}

// RequestOption customizes a single SDK call, e.g. client.Endpoints.List(ctx, WithTimeout(2*time.Second)).
type RequestOption func(*requestOptions)

// WithQuery adds/overrides a query parameter for this call.
func WithQuery(key, value string) RequestOption {
	return func(o *requestOptions) {
		if o.query == nil {
			o.query = url.Values{}
		}
		o.query.Set(key, value)
	}
}

// WithHeader adds a custom header for this call.
func WithHeader(key, value string) RequestOption {
	return func(o *requestOptions) {
		if o.headers == nil {
			o.headers = map[string]string{}
		}
		o.headers[key] = value
	}
}

// WithTimeout overrides the client's default timeout for this call only.
func WithTimeout(d time.Duration) RequestOption {
	return func(o *requestOptions) { o.timeout = d }
}

// WithMaxRetries overrides the client's default retry count for this call only.
func WithMaxRetries(n int) RequestOption {
	return func(o *requestOptions) { o.maxRetries = &n }
}

// WithIdempotencyKey sets the `idempotency_key` field RelayHub's publish-event
// endpoint accepts in its body (a real body field, not a header -- see docs/api/events.md).
func WithIdempotencyKey(key string) RequestOption {
	return func(o *requestOptions) { o.idempotencyKey = key }
}

type transport struct {
	baseURL        string
	apiKey         string
	timeout        time.Duration
	maxRetries     int
	defaultHeaders map[string]string
	httpClient     *http.Client
}

func (t *transport) do(ctx context.Context, method, path string, body any, opts ...RequestOption) (json.RawMessage, error) {
	options := &requestOptions{}
	for _, opt := range opts {
		opt(options)
	}

	maxRetries := t.maxRetries
	if options.maxRetries != nil {
		maxRetries = *options.maxRetries
	}
	timeout := t.timeout
	if options.timeout > 0 {
		timeout = options.timeout
	}

	fullURL := t.baseURL + path
	if len(options.query) > 0 {
		fullURL += "?" + options.query.Encode()
	}

	var bodyBytes []byte
	if body != nil {
		payload := body
		if options.idempotencyKey != "" {
			payload = mergeIdempotencyKey(body, options.idempotencyKey)
		}
		encoded, err := json.Marshal(payload)
		if err != nil {
			return nil, fmt.Errorf("relayhub: failed to encode request body: %w", err)
		}
		bodyBytes = encoded
	}

	var lastErr error
	for attempt := 0; attempt <= maxRetries; attempt++ {
		callCtx, cancel := context.WithTimeout(ctx, timeout)
		req, err := http.NewRequestWithContext(callCtx, method, fullURL, bytes.NewReader(bodyBytes))
		if err != nil {
			cancel()
			return nil, fmt.Errorf("relayhub: failed to build request: %w", err)
		}
		// See the matching comment in the Node SDK's transport.ts -- the backend's
		// API-key auth dependency (app/modules/api_keys/dependencies.py) reads ONLY
		// this header. Authorization: Bearer is reserved for dashboard user JWT
		// sessions, a separate auth path this client never uses.
		req.Header.Set("X-RelayHub-Api-Key", t.apiKey)
		req.Header.Set("Content-Type", "application/json")
		req.Header.Set("User-Agent", "relayhub-go/1.0.1")
		for k, v := range t.defaultHeaders {
			req.Header.Set(k, v)
		}
		for k, v := range options.headers {
			req.Header.Set(k, v)
		}

		resp, err := t.httpClient.Do(req)
		cancel()
		if err != nil {
			lastErr = &Error{Message: fmt.Sprintf("request to %s failed: %v", path, err), Status: 0}
			if attempt < maxRetries {
				time.Sleep(backoff(attempt + 1))
				continue
			}
			return nil, lastErr
		}

		respBody, readErr := io.ReadAll(resp.Body)
		resp.Body.Close()
		if readErr != nil {
			lastErr = &Error{Message: fmt.Sprintf("failed to read response from %s: %v", path, readErr), Status: 0}
			if attempt < maxRetries {
				time.Sleep(backoff(attempt + 1))
				continue
			}
			return nil, lastErr
		}

		if resp.StatusCode == 204 {
			return nil, nil
		}

		if resp.StatusCode >= 200 && resp.StatusCode < 300 {
			return respBody, nil
		}

		retryAfterSecs, hasRetryAfter := parseRetryAfter(resp.Header.Get("Retry-After"))

		if retryableStatus[resp.StatusCode] && attempt < maxRetries {
			// Retry-After, when the server sends it, REPLACES our own exponential
			// backoff for this wait -- it should never stack with it. Falling back
			// to backoff only when the server didn't tell us how long to wait.
			if hasRetryAfter {
				time.Sleep(time.Duration(retryAfterSecs * float64(time.Second)))
			} else {
				time.Sleep(backoff(attempt + 1))
			}
			continue
		}

		return nil, errorFromBody(resp.StatusCode, respBody, retryAfterSecs, hasRetryAfter)
	}

	return nil, lastErr
}

func mergeIdempotencyKey(body any, key string) map[string]any {
	encoded, err := json.Marshal(body)
	if err != nil {
		return map[string]any{"idempotency_key": key}
	}
	var m map[string]any
	if err := json.Unmarshal(encoded, &m); err != nil || m == nil {
		m = map[string]any{}
	}
	m["idempotency_key"] = key
	return m
}

func backoff(attempt int) time.Duration {
	shifted := 1 << uint(attempt-1)
	base := math.Min(float64(shifted), 8.0)
	jitter := rand.Float64() * 0.25
	return time.Duration((base + jitter) * float64(time.Second))
}

func parseRetryAfter(header string) (float64, bool) {
	if header == "" {
		return 0, false
	}
	seconds, err := strconv.ParseFloat(header, 64)
	if err != nil {
		return 0, false
	}
	return seconds, true
}

type errorEnvelope struct {
	Error struct {
		Message   string `json:"message"`
		Code      string `json:"code"`
		RequestID string `json:"request_id"`
	} `json:"error"`
}

func errorFromBody(status int, body []byte, retryAfterSecs float64, hasRetryAfter bool) *Error {
	var env errorEnvelope
	message := fmt.Sprintf("request failed with status %d", status)
	code := ""
	requestID := ""
	if err := json.Unmarshal(body, &env); err == nil && env.Error.Message != "" {
		message = env.Error.Message
		code = env.Error.Code
		requestID = env.Error.RequestID
	} else if trimmed := strings.TrimSpace(string(body)); trimmed != "" {
		message = trimmed
	}
	return &Error{
		Message:        message,
		Status:         status,
		Code:           code,
		RequestID:      requestID,
		Details:        string(body),
		RetryAfterSecs: retryAfterSecs,
		hasRetryAfter:  hasRetryAfter,
	}
}

func decode[T any](raw json.RawMessage, err error) (T, error) {
	var zero T
	if err != nil {
		return zero, err
	}
	if raw == nil {
		return zero, nil
	}
	var out T
	if unmarshalErr := json.Unmarshal(raw, &out); unmarshalErr != nil {
		return zero, fmt.Errorf("relayhub: failed to decode response: %w", unmarshalErr)
	}
	return out, nil
}

File: sdks/java/src/main/java/dev/relayhub/sdk/Transport.java
package dev.relayhub.sdk;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ThreadLocalRandom;
import java.util.stream.Collectors;

/**
 * Internal HTTP transport shared by every resource client. Not part of the
 * public API -- resource classes ({@code client.getEndpoints()}, etc.) are.
 * Handles auth headers, timeouts, exponential-backoff retries on 429/5xx and
 * connection errors, and mapping non-2xx responses to typed {@link RelayHubException} subclasses.
 */
final class Transport {
    private static final Set<Integer> RETRYABLE_STATUS = Set.of(429, 500, 502, 503, 504);

    private final String baseUrl;
    private final String apiKey;
    private final Duration timeout;
    private final int maxRetries;
    private final Map<String, String> defaultHeaders;
    private final HttpClient httpClient;
    private final ObjectMapper mapper;

    Transport(String baseUrl, String apiKey, Duration timeout, int maxRetries, Map<String, String> defaultHeaders, HttpClient httpClient, ObjectMapper mapper) {
        this.baseUrl = baseUrl;
        this.apiKey = apiKey;
        this.timeout = timeout;
        this.maxRetries = maxRetries;
        this.defaultHeaders = defaultHeaders;
        this.httpClient = httpClient;
        this.mapper = mapper;
    }

    <T> T request(String method, String path, Object body, Class<T> responseType, RequestOptions options) {
        JsonNode raw = requestRaw(method, path, body, options);
        if (raw == null) return null;
        try {
            return mapper.treeToValue(raw, responseType);
        } catch (IOException e) {
            throw new RelayHubException.ConnectionException("Failed to decode response from " + path, e);
        }
    }

    <T> T request(String method, String path, Object body, com.fasterxml.jackson.databind.JavaType responseType, RequestOptions options) {
        JsonNode raw = requestRaw(method, path, body, options);
        if (raw == null) return null;
        try {
            return mapper.convertValue(raw, responseType);
        } catch (IllegalArgumentException e) {
            throw new RelayHubException.ConnectionException("Failed to decode response from " + path, e);
        }
    }

    <T> java.util.List<T> requestList(String method, String path, Object body, Class<T> elementType, RequestOptions options) {
        com.fasterxml.jackson.databind.JavaType listType = mapper.getTypeFactory().constructCollectionType(java.util.List.class, elementType);
        java.util.List<T> result = request(method, path, body, listType, options);
        return result != null ? result : java.util.List.of();
    }

    String requestText(String method, String path, Map<String, String> query, RequestOptions options) {
        RequestOptions merged = mergeQuery(options, query);
        return requestRawText(method, path, null, merged);
    }

    private JsonNode requestRaw(String method, String path, Object body, RequestOptions options) {
        String text = requestRawText(method, path, body, options);
        if (text == null || text.isEmpty()) return null;
        try {
            return mapper.readTree(text);
        } catch (IOException e) {
            throw new RelayHubException.ConnectionException("Failed to parse response from " + path, e);
        }
    }

    private RequestOptions mergeQuery(RequestOptions options, Map<String, String> query) {
        RequestOptions.Builder b = RequestOptions.builder();
        if (options != null) {
            if (options.timeout != null) b.timeout(options.timeout);
            if (options.maxRetries != null) b.maxRetries(options.maxRetries);
            if (options.idempotencyKey != null) b.idempotencyKey(options.idempotencyKey);
            options.headers.forEach(b::header);
            options.query.forEach(b::query);
        }
        if (query != null) query.forEach((k, v) -> { if (v != null) b.query(k, v); });
        return b.build();
    }

    private String requestRawText(String method, String path, Object body, RequestOptions options) {
        String url = baseUrl + path + buildQueryString(options);
        int retries = (options != null && options.maxRetries != null) ? options.maxRetries : maxRetries;
        Duration effectiveTimeout = (options != null && options.timeout != null) ? options.timeout : timeout;

        String requestBody = null;
        if (body != null) {
            Object payload = body;
            if (options != null && options.idempotencyKey != null) {
                payload = mergeIdempotencyKey(body, options.idempotencyKey);
            }
            try {
                requestBody = mapper.writeValueAsString(payload);
            } catch (IOException e) {
                throw new RelayHubException.ConnectionException("Failed to encode request body for " + path, e);
            }
        }

        RelayHubException lastError = null;
        for (int attempt = 0; attempt <= retries; attempt++) {
            HttpRequest.Builder builder = HttpRequest.newBuilder(URI.create(url))
                    .timeout(effectiveTimeout)
                    // See the matching comment in the Node SDK's transport.ts -- the backend's
                    // API-key auth dependency (app/modules/api_keys/dependencies.py) reads ONLY
                    // this header. Authorization: Bearer is reserved for dashboard user JWT
                    // sessions, a separate auth path this client never uses.
                    .header("X-RelayHub-Api-Key", apiKey)
                    .header("Content-Type", "application/json")
                    .header("User-Agent", "relayhub-java/1.0.1");
            defaultHeaders.forEach(builder::header);
            if (options != null) options.headers.forEach(builder::header);

            HttpRequest.BodyPublisher publisher = requestBody != null
                    ? HttpRequest.BodyPublishers.ofString(requestBody)
                    : HttpRequest.BodyPublishers.noBody();
            builder.method(method, publisher);

            HttpResponse<String> response;
            try {
                response = httpClient.send(builder.build(), HttpResponse.BodyHandlers.ofString());
            } catch (IOException | InterruptedException e) {
                lastError = new RelayHubException.ConnectionException("Request to " + path + " failed: " + e.getMessage(), e);
                if (attempt < retries) { sleep(backoffMillis(attempt + 1)); continue; }
                throw lastError;
            }

            int status = response.statusCode();
            if (status == 204) return null;

            if (status >= 200 && status < 300) return response.body();

            String retryAfterHeader = response.headers().firstValue("Retry-After").orElse(null);
            Double retryAfterSeconds = null;
            if (retryAfterHeader != null) {
                try { retryAfterSeconds = Double.parseDouble(retryAfterHeader); } catch (NumberFormatException ignored) {}
            }

            if (RETRYABLE_STATUS.contains(status) && attempt < retries) {
                // Retry-After, when the server sends it, REPLACES our own exponential
                // backoff for this wait -- it should never stack with it. Falling back
                // to backoff only when the server didn't tell us how long to wait.
                sleep(retryAfterSeconds != null ? (long) (retryAfterSeconds * 1000) : backoffMillis(attempt + 1));
                continue;
            }

            throw errorFromResponse(status, response.body(), retryAfterSeconds);
        }
        throw lastError != null ? lastError : new RelayHubException.ConnectionException("Request to " + path + " failed with no response", null);
    }

    private Object mergeIdempotencyKey(Object body, String key) {
        Map<String, Object> map = mapper.convertValue(body, new com.fasterxml.jackson.core.type.TypeReference<Map<String, Object>>() {});
        Map<String, Object> merged = new HashMap<>(map);
        merged.put("idempotency_key", key);
        return merged;
    }

    private String buildQueryString(RequestOptions options) {
        if (options == null || options.query.isEmpty()) return "";
        return "?" + options.query.entrySet().stream()
                .map(e -> java.net.URLEncoder.encode(e.getKey(), java.nio.charset.StandardCharsets.UTF_8) + "="
                        + java.net.URLEncoder.encode(e.getValue(), java.nio.charset.StandardCharsets.UTF_8))
                .collect(Collectors.joining("&"));
    }

    private RelayHubException errorFromResponse(int status, String body, Double retryAfterSeconds) {
        String message = "Request failed with status " + status;
        String code = null;
        String requestId = null;
        try {
            JsonNode node = mapper.readTree(body);
            if (node.has("error")) {
                JsonNode err = node.get("error");
                if (err.has("message")) message = err.get("message").asText();
                if (err.has("code")) code = err.get("code").asText();
                if (err.has("request_id")) requestId = err.get("request_id").asText();
            }
        } catch (Exception ignored) {
            if (body != null && !body.isBlank()) message = body;
        }
        return RelayHubException.forStatus(status, message, code, requestId, body, retryAfterSeconds);
    }

    private static long backoffMillis(int attempt) {
        long base = Math.min(1000L * (1L << (attempt - 1)), 8000L);
        long jitter = ThreadLocalRandom.current().nextLong(250);
        return base + jitter;
    }

    private static void sleep(long millis) {
        try {
            Thread.sleep(millis);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }
}
